experimental/batchmap/sumdb/build/pipeline/pipeline.go

```diff
@@ -0,0 +1,140 @@
+type InputLog interface {
+	// Head returns the metadata of available entries.
+	Head() (checkpoint []byte, count int64, err error)
+	// Entries returns a PCollection of Metadata, containing entries in range [start, end).
+	Entries(s beam.Scope, start, end int64) beam.PCollection
+}
+
+// InputLogMetadata describes the provenance information of the input
+// log to be passed around atomically.
+type InputLogMetadata struct {
+	Checkpoint []byte
+	Entries    int64
+}
+
+// MapBuilder contains the static configuration for a map, and allows
+// maps at different log sizes to be built using its methods.
+type MapBuilder struct {
+	source       InputLog
+	treeID       int64
+	prefixStrata int
+	versionLogs  bool
+}
+
+// NewMapBuilder returns a MapBuilder for a map with the given configuration.
+func NewMapBuilder(source InputLog, treeID int64, prefixStrata int, versionLogs bool) MapBuilder {
+	return MapBuilder{
+		source:       source,
+		treeID:       treeID,
+		prefixStrata: prefixStrata,
+		versionLogs:  versionLogs,
+	}
+}
+
+// Create builds a map from scratch, using the first `size` entries in the
+// input log. If there aren't enough entries then it will fail.
+// It returns a PCollection of *Tile as the first output, and any logs built
+// will be output in the second PCollection (of type ModuleVersionLog).
+func (b *MapBuilder) Create(s beam.Scope, size int64) (beam.PCollection, beam.PCollection, InputLogMetadata, error) {
+	var tiles, logs beam.PCollection
+
+	endID, golden, err := b.getLogEnd(size)
+	if err != nil {
+		return tiles, logs, InputLogMetadata{}, err
+	}
+
+	records := b.source.Entries(s.Scope("source"), 0, endID)
+	entries := CreateEntries(s, b.treeID, records)
+
+	if b.versionLogs {
+		var logEntries beam.PCollection
+		logEntries, logs = MakeVersionLogs(s, b.treeID, records)
+		entries = beam.Flatten(s, entries, logEntries)
+	}
+
+	glog.Infof("Creating new map revision from range [0, %d)", endID)
+	tiles, err = batchmap.Create(s, entries, b.treeID, hash, b.prefixStrata)
+
+	return tiles, logs, InputLogMetadata{
+		Checkpoint: golden,
+		Entries:    endID,
+	}, err
+}
+
+// Update builds a map using the last version built, and updating it to
+// include all the first `size` entries from the input log. If there aren't
+// enough entries then it will fail.
+// It returns a PCollection of *Tile as the first output.
+func (b *MapBuilder) Update(s beam.Scope, lastTiles beam.PCollection, provenance InputLogMetadata, size int64) (beam.PCollection, InputLogMetadata, error) {
+	var tiles beam.PCollection
+
+	if b.versionLogs {
+		return tiles, InputLogMetadata{}, errors.New("unsupported: incremental build incompatible with version logs")
+	}
+
+	endID, golden, err := b.getLogEnd(size)
+	if err != nil {
+		return tiles, InputLogMetadata{}, err
+	}
+
+	startID := provenance.Entries
+	if startID >= endID {
+		return tiles, InputLogMetadata{}, fmt.Errorf("startID >= endID (%d > %d)", startID, endID)
```

Comment:
`>=` for names vs `>` for values in the parenthesis (how about `startID (%d) >= endID (%d)`?)

Reply:
Good catch